File: linked_list/0061_rotate_list/0061_rotate_list.py

class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def rotateRight(self, head, k):
        if not head or not head.next:
            return head

        n, cur = 1, head
        while cur.next:
            cur = cur.next
            n += 1
        cur.next = head

        cur, tail = head, cur
        for _ in xrange(n - k % n):
            tail = cur
            cur = cur.next
        tail.next = None 

        return cur 


head = ListNode(1)
head.next = ListNode(2)
head.next.next = ListNode(3)
res = Solution().rotateRight(head, 2)
print(res.val)

File: linked_list/0061_rotate_list/0061_rotate_list.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func rotateRight(head *ListNode, k int) *ListNode {
	if k == 0 || head == nil {
		return head
	}

	tail := head
	for i := 0; i < k; i++ {
		if tail.Next == nil {
			return rotateRight(head, k%(i+1))
		}
		tail = tail.Next
	}
	newTail := head
	for tail.Next != nil {
		newTail, tail = newTail.Next, tail.Next
	}

	newHead := newTail.Next
	newTail.Next, tail.Next = nil, head

	return newHead
}

func main() {
	head := &ListNode{1, &ListNode{2, &ListNode{3, nil}}}
	res := rotateRight(head, 2)
	fmt.Println(res.Val)
}
